databuilder/backends/base.py

```diff
@@ -39,67 +39,51 @@ def get_table_expression(self, table_name, schema):
         Raises:
             ValueError: If unknown table passed in
         """
-        # TODO: We currently ignore the `schema` argument here. But I think we should
-        # move towards having the supplied schema define the column types or, if not
-        # that, then we should validate that the types match.
-        return self.tables[table_name].get_expression(table_name)
+        return self.tables[table_name].get_expression(table_name, schema)
 
 
 class SQLTable:
     def learn_patient_join(self, source):
-        raise NotImplementedError()
-
-    def _make_columns(self):
-        return [
-            self._make_column(name, column) for name, column in self.columns.items()
-        ]
-
-    def _make_column(self, name, column):
-        source = column.source or name
-        type_ = TYPES_BY_NAME[column.type].value
-        sql_column = sqlalchemy.Column(source, type_, key=name)
-        return sql_column
+        self.patient_join_column = source
 
 
 class MappedTable(SQLTable):
     def __init__(self, source, columns, schema=None):
-        self.source = source
-        self.columns = columns
-        self._schema = schema
-
-    def learn_patient_join(self, source):
-        if "patient_id" not in self.columns:
-            self.columns["patient_id"] = Column("integer", source)
+        self.source_table_name = source
+        self.column_map = columns
+        # Not to be confused with the schema which defines the column names and types
+        self.db_schema_name = schema
 
-    def get_expression(self, table_name):
-        columns = self._make_columns()
-        return sqlalchemy.table(self.source, *columns, schema=self._schema)
+    def get_expression(self, table_name, schema):
+        patient_id_column = self.column_map.get("patient_id", self.patient_join_column)
```

Comment:
Related to my comment further up. Can you remind me why it's desirable to have two ways of specifying the patient_id column? (Either explicitly in the columns or via the `patient_join_column` on the backend class.)

Reply:
Well, we need the facility to specify them per-table because they aren't always consistent within a backend e.g. if a backend provider imports a third-party dataset without changing the column names — as we've seen before. But then other backend providers are careful to always name them consistently and so it's a bit annoying in that case to have to specify them over again for each table.

So we've ended up having a backend-wide default which can be overridden per-table if necessary. This doesn't seem totally unreasonable to me, though I agree it's a bit inelegant.